## 모듈 import

In [1]:
from IPython.display import Image
import numpy as np
import pandas as pd
import seaborn as sns

## 실습에 활용할 데이터셋

### 타이타닉: 탑승객의 사망자와 생존자 데이터 분석

In [ ]:
Image('https://www.gukjenews.com/news/photo/201912/1212183_991659_1921.jpg')

In [ ]:
Image('https://1.bp.blogspot.com/-vTBB13E6mtA/WzmniMnyWnI/AAAAAAACzLs/4EL_VcYCGZoO1z2QLLAIqlnO4GlYbZdZACLcBGAs/s1600/1_32.jpg')

[큰 이미지로 보기](https://static1.squarespace.com/static/5006453fe4b09ef2252ba068/t/5090b249e4b047ba54dfd258/1351660113175/TItanic-Survival-Infographic.jpg?format=1500w)

In [ ]:
Image('https://static1.squarespace.com/static/5006453fe4b09ef2252ba068/t/5090b249e4b047ba54dfd258/1351660113175/TItanic-Survival-Infographic.jpg')

건조 당시 세계 최대의 여객선이었지만,**1912년의 최초이자 최후의 항해 때 빙산과 충돌해 침몰한 비운의 여객선**. 아마도 세상에서 가장 유명한 여객선이자 침몰선일 것입니다. 

침몰한 지 100년이 넘었지만 아직까지 세계에서 가장 유명한 침몰선입니다. 

사망자 수는 1위는 아니지만, 세계적으로 유명한 영화의 영향도 있고, 당시 최첨단 기술에 대해 기대감이 컸던 사회에 큰 영향을 끼치기도 한데다가, 근대 사회에서 들어서자마자 얼마 안된, 그리고 **유명인사들이 여럿 희생된 대참사**이기 때문에 가장 유명한 침몰선이 되었습니다. 또한 이 사건을 기점으로 여러가지 안전 조약들이 생겨났으니 더더욱 그렇습니다.

In [2]:
df = sns.load_dataset("titanic")
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

### 컬럼 (column) 설명 

- survivied: 생존여부 (1: 생존, 0: 사망)
- pclass: 좌석 등급 (1등급, 2등급, 3등급)
- sex: 성별
- age: 나이
- sibsp: 형제 + 배우자 수
- parch: 부모 + 자녀 수
- fare: 좌석 요금
- embarked: 탑승 항구 (S, C, Q)
- class: pclass와 동일
- who: 성별과 동일
- adult_male: 성인 남자 여부
- deck: 데크 번호 (알파벳 + 숫자 혼용)
- embark_town: 탑승 항구 이름
- alive: 생존여부 (yes, no)
- alone: 혼자 탑승 여부

### 데이터 분석!

**주요 목표**
- Pandas를 활용하여 타이타닉호 생존자, 사망자 **데이터를 분석**합니다.
- 데이터를 토대로 생존율이 높은 승객, 생존율이 낮은 승객은 누구인지 판단합니다.

## head() 앞 부분 / tail() 뒷 부분 조회

- default 옵션 값으로 **5개의 행이 조회**됩니다.
- 괄호 안에 숫자를 넣어 명시적으로 조회하고 싶은 행의 갯수를 지정할 수 있습니다.

In [ ]:
df.head()

In [ ]:
df.tail()

In [ ]:
df.head(10)

In [ ]:
df.tail(10)

## info()

- 컬럼별 정보(information)를 보여줍니다.
- 데이터의 갯수, 그리고 데이터 타입(dtype)을 확인할 때 사용합니다.

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


**object** 타입은 쉽게 문자열이라고 생각하면 됩니다.

그런데, **category** 타입도 있습니다. category 타입은 문자열이지만, '남자' / '여자'처럼 카테고리화 할 수 있는 컬럼을 의미 합니다. 나중에 별도로 다루겠습니다.

## describe()

- 각 컬럼에 대한 요약 통계 제공
- 수치형 컬럼 (numerical column)의 통계를 기본으로 보여 줍니다.

In [ ]:
df.describe()

categorical column (문자열 컬럼)에 적용해 볼 수 없지 않습니다.

아래와 같이 `include='object'`를 통해 categorical column에 대한 요약 통계를 확인할 수 있습니다.

In [ ]:
df.describe(include='object')

## value_counts()

column 별 **값의 분포를 확인**할 때 사용합니다.

**남자, 여자, 아이의 데이터 분포를 확인**하고 싶다면 다음과 같이 실행합니다.

In [3]:
df['who'].value_counts()

who
man      537
woman    271
child     83
Name: count, dtype: int64

### 연습문제

`embark_town`은 승객의 탑승 항구를 나타내는 column 입니다. 탑승 항구별 승객 데이터 분포를 확인해 주세요.

In [9]:
# 코드를 입력해 주세요
df['embark_town'].value_counts()

embark_town
Southampton    644
Cherbourg      168
Queenstown      77
Name: count, dtype: int64

<p><strong>[출력 결과]</strong></p><pre>Southampton    644
Cherbourg      168
Queenstown      77
Name: embark_town, dtype: int64</pre>

## 속성: Attributes

속성 값은 **함수형으로 조회하지 않습니다.**

자주 활용하는 DataFrame은 **속성 값**들은 다음과 같습니다.

- ndim
- shape
- index
- columns
- values
- T

**차원**을 나타냅니다. DataFrame은 2가 출력됩니다.

In [ ]:
df.ndim

**(행, 열)** 순서로 출력됩니다.

In [ ]:
df.shape

index는 기본 설정된 **RangeIndex가 출력**됩니다.

In [ ]:
df.index

columns는 **열**을 출력 합니다. 

In [ ]:
df.columns

values는 모든 값을 출력하며, **numpy array 형식으로 출력**됩니다.

In [ ]:
df.values

T: **전치 (Transpose)** 는 Index와 Column의 축을 교환합니다.

In [ ]:
df.T

## 타입 변환 (astype)

In [ ]:
df.info()

`int32`로 변경

In [ ]:
df['pclass'].astype('int32').head()

`float32`로 변경

In [ ]:
df['pclass'].astype('float32').head()

`object`로 변경

In [ ]:
df['pclass'].astype('str').head()

`category`로 변경.

`category`로 변경시에는 Categories가 같이 출력 됩니다.

In [ ]:
df['pclass'].astype('category').head()

## 정렬 (sort)

### sort_index: index 정렬

- index 기준으로 정렬합니다. (기본 오름차순이 적용되어 있습니다.
- 내림차순 정렬을 적용하려면, `ascending=False`를 옵션 값으로 설정합니다.

In [ ]:
df.sort_index().head(5)

In [ ]:
df.sort_index(ascending=False).head(5)

### sort_values: 값에 대한 정렬

- 값을 기준으로 행을 정렬합니다.
- by에 기준이 되는 행을 설정합니다.
- by에 2개 이상의 컬럼을 지정하여 정렬할 수 있습니다.
- 오름차순/내림차순을 컬럼 별로 지정할 수 있습니다.

In [ ]:
df.sort_values(by='age').head()

내림차순 정렬: `ascending=False`

In [ ]:
df.sort_values(by='age', ascending=False).head()

**문자열 컬럼도 오름차순/내림차순 정렬이 가능**하며 알파벳 순서로 정렬됩니다.

In [ ]:
df.sort_values(by='class', ascending=False).head()

**2개 이상의 컬럼**을 기준으로 값 정렬 할 수 있습니다.

In [ ]:
df.sort_values(by=['fare', 'age']).head()

오름차순/내림차순 정렬도 컬럼 **각각에 지정**해 줄 수 있습니다.

In [ ]:
df.sort_values(by=['fare', 'age'], ascending=[False, True]).head()

### 연습문제

`tips` 데이터는 미국 레스토랑의 매출과 웨이터에게 지불한 팁을 나타내는 데이터입니다.

In [10]:
tips = sns.load_dataset('tips')
tips.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

- `total_bill`과 `tip`에 대한 내림차순 정렬을 해주세요
- 상위 10개만 출력하세요

In [14]:
# 코드를 입
tips.sort_values(by=['total_bill','tip'] , ascending=False).head(10)

total_bill    tip     sex smoker   day    time  size
170       50.81  10.00    Male    Yes   Sat  Dinner     3
212       48.33   9.00    Male     No   Sat  Dinner     4
59        48.27   6.73    Male     No   Sat  Dinner     4
156       48.17   5.00    Male     No   Sun  Dinner     6
182       45.35   3.50    Male    Yes   Sun  Dinner     3
102       44.30   2.50  Female    Yes   Sat  Dinner     3
197       43.11   5.00  Female    Yes  Thur   Lunch     4
142       41.19   5.00    Male     No  Thur   Lunch     5
184       40.55   3.00    Male    Yes   Sun  Dinner     2
95        40.17   4.73    Male    Yes   Fri  Dinner     4

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>total_bill</th>
      <th>tip</th>
      <th>sex</th>
      <th>smoker</th>
      <th>day</th>
      <th>time</th>
      <th>size</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>170</th>
      <td>50.81</td>
      <td>10.00</td>
      <td>Male</td>
      <td>Yes</td>
      <td>Sat</td>
      <td>Dinner</td>
      <td>3</td>
    </tr>
    <tr>
      <th>212</th>
      <td>48.33</td>
      <td>9.00</td>
      <td>Male</td>
      <td>No</td>
      <td>Sat</td>
      <td>Dinner</td>
      <td>4</td>
    </tr>
    <tr>
      <th>59</th>
      <td>48.27</td>
      <td>6.73</td>
      <td>Male</td>
      <td>No</td>
      <td>Sat</td>
      <td>Dinner</td>
      <td>4</td>
    </tr>
    <tr>
      <th>156</th>
      <td>48.17</td>
      <td>5.00</td>
      <td>Male</td>
      <td>No</td>
      <td>Sun</td>
      <td>Dinner</td>
      <td>6</td>
    </tr>
    <tr>
      <th>182</th>
      <td>45.35</td>
      <td>3.50</td>
      <td>Male</td>
      <td>Yes</td>
      <td>Sun</td>
      <td>Dinner</td>
      <td>3</td>
    </tr>
    <tr>
      <th>102</th>
      <td>44.30</td>
      <td>2.50</td>
      <td>Female</td>
      <td>Yes</td>
      <td>Sat</td>
      <td>Dinner</td>
      <td>3</td>
    </tr>
    <tr>
      <th>197</th>
      <td>43.11</td>
      <td>5.00</td>
      <td>Female</td>
      <td>Yes</td>
      <td>Thur</td>
      <td>Lunch</td>
      <td>4</td>
    </tr>
    <tr>
      <th>142</th>
      <td>41.19</td>
      <td>5.00</td>
      <td>Male</td>
      <td>No</td>
      <td>Thur</td>
      <td>Lunch</td>
      <td>5</td>
    </tr>
    <tr>
      <th>184</th>
      <td>40.55</td>
      <td>3.00</td>
      <td>Male</td>
      <td>Yes</td>
      <td>Sun</td>
      <td>Dinner</td>
      <td>2</td>
    </tr>
    <tr>
      <th>95</th>
      <td>40.17</td>
      <td>4.73</td>
      <td>Male</td>
      <td>Yes</td>
      <td>Fri</td>
      <td>Dinner</td>
      <td>4</td>
    </tr>
  </tbody>
</table>
</div>

- `size`를 기준으로 내림차순, `tip`을 기준으로는 오름차순 정렬을 해주세요
- 상위 10개의 데이터만 출력하세요

In [15]:
# 코드를 입력해 주세요
tips.sort_values(by=['size','tip'] , ascending=[False,True]).head(10)

total_bill   tip     sex smoker   day    time  size
125       29.80  4.20  Female     No  Thur   Lunch     6
143       27.05  5.00  Female     No  Thur   Lunch     6
156       48.17  5.00    Male     No   Sun  Dinner     6
141       34.30  6.70    Male     No  Thur   Lunch     6
187       30.46  2.00    Male    Yes   Sun  Dinner     5
216       28.15  3.00    Male    Yes   Sat  Dinner     5
142       41.19  5.00    Male     No  Thur   Lunch     5
185       20.69  5.00    Male     No   Sun  Dinner     5
155       29.85  5.14  Female     No   Sun  Dinner     5
153       24.55  2.00    Male     No   Sun  Dinner     4

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>total_bill</th>
      <th>tip</th>
      <th>sex</th>
      <th>smoker</th>
      <th>day</th>
      <th>time</th>
      <th>size</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>125</th>
      <td>29.80</td>
      <td>4.20</td>
      <td>Female</td>
      <td>No</td>
      <td>Thur</td>
      <td>Lunch</td>
      <td>6</td>
    </tr>
    <tr>
      <th>143</th>
      <td>27.05</td>
      <td>5.00</td>
      <td>Female</td>
      <td>No</td>
      <td>Thur</td>
      <td>Lunch</td>
      <td>6</td>
    </tr>
    <tr>
      <th>156</th>
      <td>48.17</td>
      <td>5.00</td>
      <td>Male</td>
      <td>No</td>
      <td>Sun</td>
      <td>Dinner</td>
      <td>6</td>
    </tr>
    <tr>
      <th>141</th>
      <td>34.30</td>
      <td>6.70</td>
      <td>Male</td>
      <td>No</td>
      <td>Thur</td>
      <td>Lunch</td>
      <td>6</td>
    </tr>
    <tr>
      <th>187</th>
      <td>30.46</td>
      <td>2.00</td>
      <td>Male</td>
      <td>Yes</td>
      <td>Sun</td>
      <td>Dinner</td>
      <td>5</td>
    </tr>
    <tr>
      <th>216</th>
      <td>28.15</td>
      <td>3.00</td>
      <td>Male</td>
      <td>Yes</td>
      <td>Sat</td>
      <td>Dinner</td>
      <td>5</td>
    </tr>
    <tr>
      <th>142</th>
      <td>41.19</td>
      <td>5.00</td>
      <td>Male</td>
      <td>No</td>
      <td>Thur</td>
      <td>Lunch</td>
      <td>5</td>
    </tr>
    <tr>
      <th>185</th>
      <td>20.69</td>
      <td>5.00</td>
      <td>Male</td>
      <td>No</td>
      <td>Sun</td>
      <td>Dinner</td>
      <td>5</td>
    </tr>
    <tr>
      <th>155</th>
      <td>29.85</td>
      <td>5.14</td>
      <td>Female</td>
      <td>No</td>
      <td>Sun</td>
      <td>Dinner</td>
      <td>5</td>
    </tr>
    <tr>
      <th>153</th>
      <td>24.55</td>
      <td>2.00</td>
      <td>Male</td>
      <td>No</td>
      <td>Sun</td>
      <td>Dinner</td>
      <td>4</td>
    </tr>
  </tbody>
</table>
</div>

## Indexing, Slicing, 조건 필터링

In [16]:
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

### loc - indexing / slicing

- indexing과 slicing을 할 수 있습니다.
- slicing은 [**시작(포함)**: **끝(포함)**] 규칙에 유의합니다. 둘 다 포함 합니다.

**indexing 예시**

In [17]:
df.loc[5, 'class']

'Third'

**fancy indexing 예시**

In [18]:
df.loc[2:5, ['age', 'fare', 'who']]

age     fare    who
2  26.0   7.9250  woman
3  35.0  53.1000  woman
4  35.0   8.0500    man
5   NaN   8.4583    man

**slicing 예시**

In [ ]:
df.loc[2:5, 'class':'deck'].head()

In [ ]:
df.loc[:6, 'class':'deck']

### loc - 조건 필터

boolean index을 만들어 조건에 맞는 데이터만 추출해 낼 수 있습니다.

In [19]:
condition = df['who'] == 'man'
condition

0       True
1      False
2      False
3      False
4       True
       ...  
886     True
887    False
888    False
889     True
890     True
Name: who, Length: 891, dtype: bool

다음 **2가지의 케이스**로 조건에 맞는 데이터만 추출 할 수 있습니다.

결과는 같습니다.

#### 케이스 1: df[condition]

In [20]:
df[condition].head()

survived  pclass   sex   age  sibsp  parch     fare embarked  class  who  \
0          0       3  male  22.0      1      0   7.2500        S  Third  man   
4          0       3  male  35.0      0      0   8.0500        S  Third  man   
5          0       3  male   NaN      0      0   8.4583        Q  Third  man   
6          0       1  male  54.0      0      0  51.8625        S  First  man   
12         0       3  male  20.0      0      0   8.0500        S  Third  man   

    adult_male deck  embark_town alive  alone  
0         True  NaN  Southampton    no  False  
4         True  NaN  Southampton    no   True  
5         True  NaN   Queenstown    no   True  
6         True    E  Southampton    no   True  
12        True  NaN  Southampton    no   True

#### 케이스 2: df.loc[condition]

In [21]:
df.loc[condition].head()

survived  pclass   sex   age  sibsp  parch     fare embarked  class  who  \
0          0       3  male  22.0      1      0   7.2500        S  Third  man   
4          0       3  male  35.0      0      0   8.0500        S  Third  man   
5          0       3  male   NaN      0      0   8.4583        Q  Third  man   
6          0       1  male  54.0      0      0  51.8625        S  First  man   
12         0       3  male  20.0      0      0   8.0500        S  Third  man   

    adult_male deck  embark_town alive  alone  
0         True  NaN  Southampton    no  False  
4         True  NaN  Southampton    no   True  
5         True  NaN   Queenstown    no   True  
6         True    E  Southampton    no   True  
12        True  NaN  Southampton    no   True

다만, **loc를 사용하는 것을 추천**합니다. (값 대입시 issue 발생)

In [22]:
df[condition]['age']

0      22.0
4      35.0
5       NaN
6      54.0
12     20.0
       ... 
883    28.0
884    25.0
886    27.0
889    26.0
890    32.0
Name: age, Length: 537, dtype: float64

In [23]:
df[condition]['age'] = 2

C:\Users\dndjf\AppData\Local\Temp\ipykernel_29324\4257232175.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[condition]['age'] = 2


다음과 같은 **경고 창**이 뜹니다.

/home/ubuntu/anaconda3/envs/tensorflow2_p36/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: <br>
A value is trying to be set on a copy of a slice from a DataFrame.<br>
**Try using .loc[row_indexer,col_indexer] = value instead**<br>
<br>
See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy<br>
  """Entry point for launching an IPython kernel.<br>

값을 대입하여 변경했음에도 불구하고 **값이 변경 되지 않습니다.**

In [ ]:
df[condition]['age']

`loc`를 사용하면 이러한 문제가 발생하지 않아 **실수를 줄일** 수 있습니다.

In [ ]:
df.loc[condition, 'age'] = 10

In [ ]:
df[condition].head()

### loc - 다중 조건

다중 조건은 먼저 condition을 정의하고 **&** 와 **|** 연산자로 **복합 조건을 생성**합니다.

In [ ]:
# 조건1 정의
condition1 = (df['fare'] > 30)

# 조건2 정의
condition2 = (df['who'] == 'woman')

In [ ]:
df.loc[condition1 & condition2]

In [ ]:
df.loc[condition1 | condition2]

### 연습문제

데이터를 다시 로드 합니다.

In [ ]:
df = sns.load_dataset("titanic")
df.head()

다음 조건을 만족하는 코드를 입력하세요.

- 나이가 30살 이상 남자 승객 조건 필터링
- `fare`를 많이 낸 순서로 내림차순 정렬
- 상위 10개를 출력

In [33]:
# 코드를 입력해 주세요
condition1=(df['age'] >= 30)
condition2=(df['sex'] == 'male')

In [35]:
df[condition1 & condition2].sort_values(by=['fare'],ascending=False).head(10)

survived  pclass   sex   age  sibsp  parch      fare embarked  class  \
679         1       1  male  36.0      0      1  512.3292        C  First   
737         1       1  male  35.0      0      0  512.3292        C  First   
438         0       1  male  64.0      1      4  263.0000        S  First   
332         0       1  male  38.0      0      1  153.4625        S  First   
660         1       1  male  50.0      2      0  133.6500        S  First   
390         1       1  male  36.0      1      2  120.0000        S  First   
659         0       1  male  58.0      0      2  113.2750        C  First   
698         0       1  male  49.0      1      1  110.8833        C  First   
544         0       1  male  50.0      1      0  106.4250        C  First   
224         1       1  male  38.0      1      0   90.0000        S  First   

     who  adult_male deck  embark_town alive  alone  
679  man        True    B    Cherbourg   yes  False  
737  man        True    B    Cherbourg   yes   True  
438  man        True    C  Southampton    no  False  
332  man        True    C  Southampton    no  False  
660  man        True  NaN  Southampton   yes  False  
390  man        True    B  Southampton   yes  False  
659  man        True    D    Cherbourg    no  False  
698  man        True    C    Cherbourg    no  False  
544  man        True    C    Cherbourg    no  False  
224  man        True    C  Southampton   yes  False

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>survived</th>
      <th>pclass</th>
      <th>sex</th>
      <th>age</th>
      <th>sibsp</th>
      <th>parch</th>
      <th>fare</th>
      <th>embarked</th>
      <th>class</th>
      <th>who</th>
      <th>adult_male</th>
      <th>deck</th>
      <th>embark_town</th>
      <th>alive</th>
      <th>alone</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>679</th>
      <td>1</td>
      <td>1</td>
      <td>male</td>
      <td>36.0</td>
      <td>0</td>
      <td>1</td>
      <td>512.3292</td>
      <td>C</td>
      <td>First</td>
      <td>man</td>
      <td>True</td>
      <td>B</td>
      <td>Cherbourg</td>
      <td>yes</td>
      <td>False</td>
    </tr>
    <tr>
      <th>737</th>
      <td>1</td>
      <td>1</td>
      <td>male</td>
      <td>35.0</td>
      <td>0</td>
      <td>0</td>
      <td>512.3292</td>
      <td>C</td>
      <td>First</td>
      <td>man</td>
      <td>True</td>
      <td>B</td>
      <td>Cherbourg</td>
      <td>yes</td>
      <td>True</td>
    </tr>
    <tr>
      <th>438</th>
      <td>0</td>
      <td>1</td>
      <td>male</td>
      <td>64.0</td>
      <td>1</td>
      <td>4</td>
      <td>263.0000</td>
      <td>S</td>
      <td>First</td>
      <td>man</td>
      <td>True</td>
      <td>C</td>
      <td>Southampton</td>
      <td>no</td>
      <td>False</td>
    </tr>
    <tr>
      <th>332</th>
      <td>0</td>
      <td>1</td>
      <td>male</td>
      <td>38.0</td>
      <td>0</td>
      <td>1</td>
      <td>153.4625</td>
      <td>S</td>
      <td>First</td>
      <td>man</td>
      <td>True</td>
      <td>C</td>
      <td>Southampton</td>
      <td>no</td>
      <td>False</td>
    </tr>
    <tr>
      <th>660</th>
      <td>1</td>
      <td>1</td>
      <td>male</td>
      <td>50.0</td>
      <td>2</td>
      <td>0</td>
      <td>133.6500</td>
      <td>S</td>
      <td>First</td>
      <td>man</td>
      <td>True</td>
      <td>NaN</td>
      <td>Southampton</td>
      <td>yes</td>
      <td>False</td>
    </tr>
    <tr>
      <th>390</th>
      <td>1</td>
      <td>1</td>
      <td>male</td>
      <td>36.0</td>
      <td>1</td>
      <td>2</td>
      <td>120.0000</td>
      <td>S</td>
      <td>First</td>
      <td>man</td>
      <td>True</td>
      <td>B</td>
      <td>Southampton</td>
      <td>yes</td>
      <td>False</td>
    </tr>
    <tr>
      <th>659</th>
      <td>0</td>
      <td>1</td>
      <td>male</td>
      <td>58.0</td>
      <td>0</td>
      <td>2</td>
      <td>113.2750</td>
      <td>C</td>
      <td>First</td>
      <td>man</td>
      <td>True</td>
      <td>D</td>
      <td>Cherbourg</td>
      <td>no</td>
      <td>False</td>
    </tr>
    <tr>
      <th>698</th>
      <td>0</td>
      <td>1</td>
      <td>male</td>
      <td>49.0</td>
      <td>1</td>
      <td>1</td>
      <td>110.8833</td>
      <td>C</td>
      <td>First</td>
      <td>man</td>
      <td>True</td>
      <td>C</td>
      <td>Cherbourg</td>
      <td>no</td>
      <td>False</td>
    </tr>
    <tr>
      <th>544</th>
      <td>0</td>
      <td>1</td>
      <td>male</td>
      <td>50.0</td>
      <td>1</td>
      <td>0</td>
      <td>106.4250</td>
      <td>C</td>
      <td>First</td>
      <td>man</td>
      <td>True</td>
      <td>C</td>
      <td>Cherbourg</td>
      <td>no</td>
      <td>False</td>
    </tr>
    <tr>
      <th>224</th>
      <td>1</td>
      <td>1</td>
      <td>male</td>
      <td>38.0</td>
      <td>1</td>
      <td>0</td>
      <td>90.0000</td>
      <td>S</td>
      <td>First</td>
      <td>man</td>
      <td>True</td>
      <td>C</td>
      <td>Southampton</td>
      <td>yes</td>
      <td>False</td>
    </tr>
  </tbody>
</table>
</div>

다음 조건을 만족하는 코드를 입력하세요.

- 나이가 20살 이상 40살 미만인 승객
- `pclass`가 1등급 혹은 2등급인 승객
- 열(column)은 `survived`, `pclass`, `age`, `fare` 만 나오게 출력
- 10개만 출력

In [48]:
# 코드를 입력해 주세요
#condition1=(df['age'] >= 20 )
#condition2=(df['pclass'] == '1')
c1=(df['age'] >= 20) & (df['age'] < 40 )
#condition3=(df['age'] < 40 )
#condition2=(df['pclass'] == '2')
c2=(df['pclass'] == 1) or (df['pclass'] == 2)
df.loc[c1 & c2]

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>survived</th>
      <th>pclass</th>
      <th>age</th>
      <th>fare</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>1</td>
      <td>38.0</td>
      <td>71.2833</td>
    </tr>
    <tr>
      <th>3</th>
      <td>1</td>
      <td>1</td>
      <td>35.0</td>
      <td>53.1000</td>
    </tr>
    <tr>
      <th>20</th>
      <td>0</td>
      <td>2</td>
      <td>35.0</td>
      <td>26.0000</td>
    </tr>
    <tr>
      <th>21</th>
      <td>1</td>
      <td>2</td>
      <td>34.0</td>
      <td>13.0000</td>
    </tr>
    <tr>
      <th>23</th>
      <td>1</td>
      <td>1</td>
      <td>28.0</td>
      <td>35.5000</td>
    </tr>
    <tr>
      <th>34</th>
      <td>0</td>
      <td>1</td>
      <td>28.0</td>
      <td>82.1708</td>
    </tr>
    <tr>
      <th>41</th>
      <td>0</td>
      <td>2</td>
      <td>27.0</td>
      <td>21.0000</td>
    </tr>
    <tr>
      <th>53</th>
      <td>1</td>
      <td>2</td>
      <td>29.0</td>
      <td>26.0000</td>
    </tr>
    <tr>
      <th>56</th>
      <td>1</td>
      <td>2</td>
      <td>21.0</td>
      <td>10.5000</td>
    </tr>
    <tr>
      <th>61</th>
      <td>1</td>
      <td>1</td>
      <td>38.0</td>
      <td>80.0000</td>
    </tr>
  </tbody>
</table>
</div>

다음 조건을 만족하는 코드를 입력하세요

- `pclass`가 1, 2등급인 승객 중 나이 값이 없는(NaN) 데이터는 나이의 평균 값으로 채워주세요

In [50]:
age_mean=df['age'].mean()
# 코드를 입력해 주세요
c1=df['pclass'] <3
c2=df['age'].isnull()

df.loc[c1&c2,'age']=age_mean

In [51]:
# 검증코드
print(df['age'].isnull().sum())
print(df['age'].mean())

136
29.69911764705882


<p><strong>[출력 결과]</strong></p><pre>136
29.69911764705882
</pre>

## iloc

- `loc`와 유사하지만, index만 허용합니다.
- loc와 마찬가지고, indexing / slicing 모두 가능합니다.

In [ ]:
df.head()

### indexing

In [ ]:
df.iloc[1, 3]

### Fancy Indexing

In [ ]:
df.iloc[[0, 3, 4], [0, 1, 5, 6]]

### Slicing

In [ ]:
df.iloc[:3, :5]

## at

하나의 인덱스만 가져옵니다. `loc`보다 속도가 빠르다는 장점은 있지만, 실질적인 효용성은 떨어집니다. 그냥 `loc`를 사용해도 똑같은 결과를 얻을 수 있습니다.

In [ ]:
%timeit df.loc[0, 'fare']

In [ ]:
%timeit df.at[0, 'fare']

## iat

하나의 인덱스만 가져옵니다. 속도가 빠르다는 장점은 있지만, 1개의 데이터만 조회 가능합니다. `iloc`로 대체 사용가능합니다.

In [ ]:
%timeit df.iloc[0, 5]

In [ ]:
%timeit df.iat[0, 5]

## where

[도큐먼트](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.where.html)

`DataFrame.where(cond, other=nan, inplace=False, axis=None, level=None, errors='raise', try_cast=False)`

Pandas의 `where`는 Numpy의 `where`와 동작이 다릅니다.

- cond: True/False로 판단될 수 있는 식
- other: condition을 만족하지 못하는 요소에 할당 할 값

In [ ]:
df.tail(5)

**fare 컬럼**에 적용하는 예시

In [52]:
df['fare'].where(df['fare'] < 20, 0).tail(10)

881     7.8958
882    10.5167
883    10.5000
884     7.0500
885     0.0000
886    13.0000
887     0.0000
888     0.0000
889     0.0000
890     7.7500
Name: fare, dtype: float64

전체 DataFrame에 적용시 다음과 같이 조건이 해당하는 값 외에는 모두 NaN 값으로 채워지게 됩니다.

In [53]:
df.where(df['fare'] < 10)

survived  pclass     sex   age  sibsp  parch   fare embarked  class  \
0         0.0     3.0    male  22.0    1.0    0.0  7.250        S  Third   
1         NaN     NaN     NaN   NaN    NaN    NaN    NaN      NaN    NaN   
2         1.0     3.0  female  26.0    0.0    0.0  7.925        S  Third   
3         NaN     NaN     NaN   NaN    NaN    NaN    NaN      NaN    NaN   
4         0.0     3.0    male  35.0    0.0    0.0  8.050        S  Third   
..        ...     ...     ...   ...    ...    ...    ...      ...    ...   
886       NaN     NaN     NaN   NaN    NaN    NaN    NaN      NaN    NaN   
887       NaN     NaN     NaN   NaN    NaN    NaN    NaN      NaN    NaN   
888       NaN     NaN     NaN   NaN    NaN    NaN    NaN      NaN    NaN   
889       NaN     NaN     NaN   NaN    NaN    NaN    NaN      NaN    NaN   
890       0.0     3.0    male  32.0    0.0    0.0  7.750        Q  Third   

       who adult_male deck  embark_town alive  alone  
0      man       True  NaN  Southampton    no  False  
1      NaN        NaN  NaN          NaN   NaN    NaN  
2    woman      False  NaN  Southampton   yes   True  
3      NaN        NaN  NaN          NaN   NaN    NaN  
4      man       True  NaN  Southampton    no   True  
..     ...        ...  ...          ...   ...    ...  
886    NaN        NaN  NaN          NaN   NaN    NaN  
887    NaN        NaN  NaN          NaN   NaN    NaN  
888    NaN        NaN  NaN          NaN   NaN    NaN  
889    NaN        NaN  NaN          NaN   NaN    NaN  
890    man       True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

## isin

특정 값의 포함 여부는 isin 함수를 통해 비교가 가능합니다. (파이썬의 in 키워드는 사용 불가 합니다.)

In [ ]:
sample = pd.DataFrame({'name': ['kim', 'lee', 'park', 'choi'], 
                        'age': [24, 27, 34, 19]
                      })
sample

In [ ]:
sample['name'].isin(['kim', 'lee'])

In [ ]:
sample.isin(['kim', 'lee'])

`loc`를 활용한 **조건 필터링**으로도 찰떡궁합입니다.

In [ ]:
condition = sample['name'].isin(['kim', 'lee'])

In [ ]:
sample.loc[condition]

### 연습문제

In [54]:
tips = sns.load_dataset('tips')
tips.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

- `tips` 데이터셋 중 `day`가 **금요일(Fri), 토요일(Sat)** 만 필터링 합니다.
- `tip`이 $10보다 적게 낸 데이터만 필터링합니다.
- 컬럼은 `total_bill`, `tip`, `smoker`, `time`만 출력합니다.
- 상위 10개 행만 출력합니다.

In [62]:
# 코드를 입력해 주세요
c1=tips['day'].isin(['Fri','Sat'])
c2=tips['tip']<10

tips.loc[c1&c2][['total_bill','tip','smoker','time']].head(10)

total_bill   tip smoker    time
19       20.65  3.35     No  Dinner
20       17.92  4.08     No  Dinner
21       20.29  2.75     No  Dinner
22       15.77  2.23     No  Dinner
23       39.42  7.58     No  Dinner
24       19.82  3.18     No  Dinner
25       17.81  2.34     No  Dinner
26       13.37  2.00     No  Dinner
27       12.69  2.00     No  Dinner
28       21.70  4.30     No  Dinner

<p><strong>[출력 결과]</strong></p><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>total_bill</th>
      <th>tip</th>
      <th>smoker</th>
      <th>time</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>19</th>
      <td>20.65</td>
      <td>3.35</td>
      <td>No</td>
      <td>Dinner</td>
    </tr>
    <tr>
      <th>20</th>
      <td>17.92</td>
      <td>4.08</td>
      <td>No</td>
      <td>Dinner</td>
    </tr>
    <tr>
      <th>21</th>
      <td>20.29</td>
      <td>2.75</td>
      <td>No</td>
      <td>Dinner</td>
    </tr>
    <tr>
      <th>22</th>
      <td>15.77</td>
      <td>2.23</td>
      <td>No</td>
      <td>Dinner</td>
    </tr>
    <tr>
      <th>23</th>
      <td>39.42</td>
      <td>7.58</td>
      <td>No</td>
      <td>Dinner</td>
    </tr>
    <tr>
      <th>24</th>
      <td>19.82</td>
      <td>3.18</td>
      <td>No</td>
      <td>Dinner</td>
    </tr>
    <tr>
      <th>25</th>
      <td>17.81</td>
      <td>2.34</td>
      <td>No</td>
      <td>Dinner</td>
    </tr>
    <tr>
      <th>26</th>
      <td>13.37</td>
      <td>2.00</td>
      <td>No</td>
      <td>Dinner</td>
    </tr>
    <tr>
      <th>27</th>
      <td>12.69</td>
      <td>2.00</td>
      <td>No</td>
      <td>Dinner</td>
    </tr>
    <tr>
      <th>28</th>
      <td>21.70</td>
      <td>4.30</td>
      <td>No</td>
      <td>Dinner</td>
    </tr>
  </tbody>
</table>
</div>